### ISSUE TITLE
Any with precompiled matcher

### ISSUE THREAD

**Issue by mathieumg** [micromatch/micromatch#35]
Hi, great project!

Is there any way to compare against multiple globs (like in https://github.com/jonschlinkert/micromatch#any) but using a precompiled matcher (like in https://github.com/jonschlinkert/micromatch#matcher)?

Thank you!

**Comment by jonschlinkert** [comment]
> Is there any way to compare against multiple globs

Agreed! seems like I've probably created what you describe on my own projects half a dozen times... this should definitely be added to the API. 

thanks for the kind words

**Comment by mathieumg** [comment]
:grin: Cool, thanks for the quick reply!

**Comment by julien-f** [comment]
My basic implementation which can serve as a base https://gist.github.com/julien-f/5b9a3537eb82a34b04e2 :)

**Comment by jonschlinkert** [comment]
related https://github.com/jonschlinkert/micromatch/issues/48

**Comment by jonschlinkert** [comment]
@julien-f thanks!

**Comment by ngryman** [comment]
@jonschlinkert Do you plan to add it? I need this too 👍

**Comment by loganfsmyth** [comment]
I just made the mistake of assuming that

    items.filter(micromatch.filter(patterns))

would be the same as

    micromatch(items, patterns)

am I right that this issue is aimed at resolving that type of thing?

**Comment by jonschlinkert** [comment]
@loganfsmyth I *think* so. But I'd love some feedback on this. 

In the [upcoming release](https://github.com/jonschlinkert/micromatch/pull/76), I'm planning on adding support for multiple globs to  `.matcher`, `.any` and `.contains`. (I generalized a "compose" function that is used in [.matcher](https://github.com/jonschlinkert/micromatch/blob/dev/lib/utils.js#L41) in the dev branch. The plan is to use that same utility in `.contains` and `.any`. 

However.. I'm also planning on removing `.filter`, as it seems like there are other more effective or more specific methods to accomplish the same thing. 

Does this answer your questions? Again any feedback is welcome

**Comment by loganfsmyth** [comment]
I don't think I have enough background to answer all of this. Essentially I was trying to implement a `.gitignore`-style thing and originally wanted to cache a `filter` function per-`ignore` file, but since `.filter` wanted to match _all_ of the patterns, I was unable to use it, assuming I understood the issues I was seeing well enough.

**Comment by jonschlinkert** [comment]
closing, since this is resolved in the 3.0 release. `.isMatch` leverages memoization, so the result should be the same as pre-compiling the matcher, but with the added advantage of lazily compiling the regex the first time the matcher is actually called. If there is something more specific you'd like to accomplish with `.any`, please feel free to create another issue, I'd be happy to discuss. 

@loganfsmyth I know this has been a while, but after re-reading your first comment, I think I misread part of the example and misunderstood. The `.matcher` method returns a memoized function that can be passed to `Array.filter`. For example:

```js
var mm = require('micromatch');
var arr = ['foo', 'bar', 'baz'];
console.log(arr.filter(mm.matcher('b*')));
```

However, if you were planning on filtering using actual `.gitignore` patterns, and you want the results to be the same as what they would be in git, there would be more to it than that. In git, wildmatch is used for matching, which follows different rules in some cases than a bash-style matcher (like micromatch). In particular, wildmatch allows for re-inclusion of previously excluded paths. I'd be happy to discuss on another issue if this is still up in the air.
